components/taxonomy/src/adapters/adapter.rs

```diff
@@ -0,0 +1,364 @@
+use misc::util::Description;
+use io::types::*;
+
+use transformable_channels::mpsc::*;
+
+use std::sync::Arc;
+
+pub type PerFeature<T> = Vec<(Id<FeatureId>, T)>;
+pub type PerFeatureResult<T> = PerFeature<Result<T, Error>>;
+
+/// Description of a service.
+///
+/// Typically, a service represents a physical device, e.g. a light, a heater, a doorlock, etc.
+///
+/// Services typically implement several `Feature`s.
+#[derive(Clone)]
+pub struct Service {
+    /// Tags describing the service.
+    ///
+    /// These tags can be set by the user, adapters or
+    /// applications. They are used by applications to find services and
+    /// services.
+    ///
+    /// For instance, a user may set tag "entrance" to all services
+    /// placed in the entrance of his house, or a tag "blue" to a service
+    /// controlling blue lights. An adapter may set tags "plugged" or
+    /// "battery" to devices that respectively depend on a plugged
+    /// power source or on a battery.
+    pub tags: Vec<Id<TagId>>,
+
+    /// An id unique to this service.
+    pub id: Id<ServiceId>,
+
+    /// Service properties that are set at creation time.
+    /// For instance, these can be device manufacturer, model, etc.
+    ///
+    /// By opposition to tags, these cannot be used by applications to discover a service, but
+    /// they can be used by applications to customize the experience. For instance, if a service
+    /// represents a device manufactured by Initrode, it can offer a link to Initrode's customer
+    /// support as a property.
+    pub properties: Vec<(String, String)>,
+
+    /// Identifier of the adapter for this service.
+    pub adapter: Id<AdapterId>,
+}
+
+/// Description of a feature of a service.
+///
+/// Typically, a service implements several `Feature`s. For instance, a service dealing with lights
+/// may implement:
+///
+/// - a `Feature` for turning the lights on and off;
+/// - a `Feature` for controlling luminosity;
+/// - a `Feature` for controlling color;
+/// - ...
+///
+/// Features are attached to a `Type`, which specifies the values that can be sent to the feature
+/// (if the feature supports SEND operations), fetched or watched from the feature (if the feature
+/// supports READ/WATCH operations).
+///
+/// # Example
+///
+/// ```
+/// use foxbox_taxonomy::adapters::adapter::*;
+/// use foxbox_taxonomy::misc::util::*;
+/// use foxbox_taxonomy::io::serialize::*;
+/// use foxbox_taxonomy::io::types::*;
+/// use foxbox_taxonomy::library::*;
+/// use std::sync::Arc;
+///
+/// // A Featured used to turn lights on or off.
+/// let turn_light_on = Feature {
+///   implements: vec!["light/on", "light/x-on"],
+///
+///   // We support sending on or off.
+///   send: Some(Signature {
+///     accepts: Expects::Requires(Arc::new(IsOnFormat)),
+///     ..Signature::default()
+///   }),
+///
+///   .. Feature::empty(&Id::new("light #1"), &Id::new("lights manager #1"))
+/// };
+/// ```
+#[derive(Clone)]
+pub struct Feature<'a> {
```

Comment:
I think we need something to indicate a feature is user-facing or configuration stuff (eg: shoud the temperature be in C or F, or should the polling happen once every minute, or once every 10 minutes, etc)